security/pkg/nodeagent/monitoring/monitoring.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package monitoring
+
+import "github.com/prometheus/client_golang/prometheus"
+
+const (
+	// Namespace of the metrics in the Prometheus system. This is not the namespace where Citadel agents
+	// reside, e.g. istio-system.
+	citadelAgentMetricsNamespace = "citadel_agent"
+)
+
+var (
+	totalOutgoingRequests = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: citadelAgentMetricsNamespace,
+		Name:      "total_outgoing_requests",
+		Help:      "the total requests from Citadel agent to the token exchange service or to the CA",
+	}, []string{"request_type"})
+
+	totalSuccessfulOutgoingRequests = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: citadelAgentMetricsNamespace,
+		Name:      "total_outgoing_successful_requests",
+		Help: "the total successful requests from Citadel agent to the token exchange service or" +
+			" to the CA",
+	}, []string{"request_type"})
+
+	numRetries = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: citadelAgentMetricsNamespace,
+		Name:      "num_retries",
+		Help: "numRetries is the number of retries from Citadel agent to the token exchange service" +
+			" or to the CA from each request.",
+	}, []string{"request_type"})
+
+	outgoingLatency = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: citadelAgentMetricsNamespace,
+		Name:      "latency",
+		Help:      "the round-trip time from Citadel agent to the token exchange service or to the CA",
```

Comment:
how about expose 3 metrics for latency ?
1. duration [start calling token exchange, getting result(success/fail)]
2. duration  [start calling CA, getting result(success/fail)]
3. duration [receive envoy request, push down cert]

Reply:
I think we should separate the latency to find the culprits easily. If we want to calculate the whole trip latency, we can easily add them. Wdyt? This PR will focus on latency from Citadel agent to token exchange server and the CA. Latency with Envoy sidecar will be in a separate PR.